plugins/sonarqube-backend/src/service/sonarqubeInfoProvider.ts

```diff
@@ -0,0 +1,371 @@
+      );
+    }
+    return instanceConfig;
+  }
+}
+
+/**
+ * @public
+ *
+ * Use default config and annotations, build using fromConfig static function.
+ */
+export class DefaultSonarqubeInfoProvider implements SonarqubeInfoProvider {
+  private constructor(private readonly config: SonarqubeConfig) {}
+
+  /**
+   * Generate an instance from a Config instance
+   * @param config - Backend configuration
+   */
+  static fromConfig(config: Config): DefaultSonarqubeInfoProvider {
+    return new DefaultSonarqubeInfoProvider(SonarqubeConfig.fromConfig(config));
+  }
+
+  /**
+   * Retrieve all supported metrics from a sonarqube instance.
+   *
+   * @param instanceUrl - URL of the sonarqube instance
+   * @param token - token to access the sonarqube instance
+   * @returns The list of supported metrics, if no metrics are supported an empty list is provided in the promise
+   * @private
+   */
+  private static async getSupportedMetrics(
+    instanceUrl: string,
+    token: string,
+  ): Promise<string[]> {
+    const metrics: string[] = [];
+    let nextPage: number = 1;
+
+    for (;;) {
+      const result = await DefaultSonarqubeInfoProvider.callApi<{
+        metrics: Array<{ key: string }>;
+        total: number;
+      }>(instanceUrl, 'api/metrics/search', token, { ps: 500, p: nextPage });
+      metrics.push(...(result?.metrics?.map(m => m.key) ?? []));
+
+      if (result && metrics.length < result.total) {
+        nextPage++;
+        continue;
+      }
+
+      return metrics;
+    }
+  }
+
+  /**
+   * Call an API with provided arguments
+   * @param url - URL of the API to call
+   * @param path - path to call
+   * @param authToken - token used as basic auth user without password
+   * @param query - parameters to provide to the call
+   * @returns A promise on the answer to the API call if the answer status code is 200, undefined otherwise.
+   * @private
+   */
+  private static async callApi<T>(
+    url: string,
+    path: string,
+    authToken: string,
+    query: { [key in string]: any },
+  ): Promise<T | undefined> {
+    // Sonarqube auth use basic with token as username and no password
+    // but standard dictate the colon (separator) need to stay here despite the
+    // lack of password
+    const encodedAuthToken = Buffer.from(`${authToken}:`).toString('base64');
+
+    const response = await fetch(
+      `${url}/${path}?${new URLSearchParams(query).toString()}`,
+      {
+        headers: {
+          'Content-Type': 'application/json',
+          Authorization: `Basic ${encodedAuthToken}`,
+        },
+      },
+    );
+    if (response.status === 200) {
+      return (await response.json()) as T;
+    }
+    return undefined;
+  }
+
+  /**
+   * {@inheritDoc SonarqubeInfoProvider.getBaseUrl}
+   * @throws Error If configuration can't be retrieved.
+   */
+  getBaseUrl({ instanceName }: { instanceName: string }): { baseUrl: string } {
```

Comment:
Here's where we could make `instanceName` optional as we already default it to empty?

Reply:
Or actually making it optional in `getInstanceConfig` is fine there too actually. Don't need to default to empty strings.

Reply:
Empty strings are a little bit of a smell for me.